***Name:*** TSION LEGESSE  
***Date:*** 4/1/2026  
***Student ID:*** GSR/1433/18

***ADVANCED PROGRAMMING ASSIGNMENT 4***

In [3]:
pip install biopython

In [4]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO

print("Biopython core modules imported successfully.")

Biopython core modules imported successfully.


# ***Exercise 1 — Write a GC Content Function***

In [5]:
# Exercise 1: write your solution here

def calculate_gc(sequence):
    # Count G and C bases
    gc_count = sequence.count("G") + sequence.count("C")
    
    # Calculate GC percentage
    gc_percentage = (gc_count / len(sequence)) * 100
    
    return gc_percentage

# Test your function
print(calculate_gc("ATGC"))

50.0


# ***Exercise 2 — Translate Your Own DNA Sequence***

Create a DNA sequence that starts with ATG, translate it, and explain the resulting amino acid sequence.

In [6]:
from Bio.Seq import Seq

# Exercise 2: write your code here

student_dna = Seq("ATGGCCATT")

# Translate DNA sequence into amino acids
protein = student_dna.translate()

print("DNA Sequence:", student_dna)
print("Amino Acid Sequence:", protein)

DNA Sequence: ATGGCCATT
Amino Acid Sequence: MAI


# ***Exercise 3 — Create a SeqRecord***

Create your own SeqRecord with:

an ID

a name

a description

a DNA sequence

In [7]:
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord

# Exercise 3 — Create a SeqRecord

my_record = SeqRecord(
    Seq("ATGCGTACGTTAG"),
    id="SEQ001",
    name="StudentSequence",
    description="Example DNA sequence created for Exercise 3"
)

# Print the SeqRecord
print(my_record)

ID: SEQ001
Name: StudentSequence
Description: Example DNA sequence created for Exercise 3
Number of features: 0
Seq('ATGCGTACGTTAG')


# ***Exercise 4 — FASTA Summary Table***

Read the example FASTA file and create a summary table with:

sequence ID
sequence length
GC content
number of A, T, G, and C bases

In [9]:
from Bio import SeqIO

# Function to calculate GC content
def calculate_gc(sequence):
    sequence = str(sequence)

    gc_count = sequence.count("G") + sequence.count("C")

    return (gc_count / len(sequence)) * 100


# Create FASTA file
fasta_text = """>seq1 Example sequence 1
ATGCGTACGTAGCTAGCTAG
>seq2 Example sequence 2
ATGCCCGGGTTTAAACCCGGG
>seq3 Example sequence 3
ATATATATATATATATATATA
"""

with open("example_sequences.fasta", "w") as handle:
    handle.write(fasta_text)

print("Example FASTA file created.")


# Read FASTA file
for record in SeqIO.parse("example_sequences.fasta", "fasta"):

    print("ID:", record.id)
    print("Description:", record.description)
    print("Sequence:", record.seq)
    print("Length:", len(record.seq))

    print("GC content:", f"{calculate_gc(record.seq):.2f}%")

    # Count bases
    sequence = str(record.seq)

    print("A count:", sequence.count("A"))
    print("T count:", sequence.count("T"))
    print("G count:", sequence.count("G"))
    print("C count:", sequence.count("C"))

    print("-" * 40)

Example FASTA file created.
ID: seq1
Description: seq1 Example sequence 1
Sequence: ATGCGTACGTAGCTAGCTAG
Length: 20
GC content: 50.00%
A count: 5
T count: 5
G count: 6
C count: 4
----------------------------------------
ID: seq2
Description: seq2 Example sequence 2
Sequence: ATGCCCGGGTTTAAACCCGGG
Length: 21
GC content: 61.90%
A count: 4
T count: 4
G count: 7
C count: 6
----------------------------------------
ID: seq3
Description: seq3 Example sequence 3
Sequence: ATATATATATATATATATATA
Length: 21
GC content: 0.00%
A count: 11
T count: 10
G count: 0
C count: 0
----------------------------------------


# ***Exercise 5 — Discussion***

# Why is GenBank More Informative than FASTA for Genome Annotation Analysis?

GenBank records are more informative than FASTA files because they contain not only the biological sequence, but also detailed annotations and metadata that describe the biological meaning of the sequence.

FASTA files mainly store raw nucleotide or protein sequences with minimal description, while GenBank files provide structured information useful for genome annotation and bioinformatics analysis.

| Feature | FASTA | GenBank |
|---|---|---|
| Sequence data | Yes | Yes |
| Gene annotations | No | Yes |
| CDS regions | No | Yes |
| Protein products | No | Yes |
| Organism metadata | Limited | Detailed |
| Scientific references | No | Yes |

---

## 1. Gene Names

FASTA files usually contain only a short header line:

```text
>seq1
ATGCGTACGTAG
gene            100..900
                /gene="cytb"
                2. CDS Coordinates

GenBank files contain CDS (Coding Sequence) coordinates that indicate where protein-coding regions start and end.
CDS             120..850
These coordinates help researchers:

extract coding regions
identify exons
analyze mutations
translate DNA into proteins

FASTA files do not contain positional information.
3. Protein Products

GenBank records often describe the protein encoded by the gene.

Example:

/product="cytochrome b"

This annotation explains the biological product generated from the sequence.

FASTA files only provide sequence information and do not explain what protein or function is associated with the sequence.
4. Organism Metadata

GenBank includes detailed metadata about the organism:

ORGANISM  Bos taurus

Additional metadata may include:

taxonomy
strain
tissue source
chromosome number
mitochondrial or nuclear origin

This information is essential for:

evolutionary studies
comparative genomics
phylogenetic analysis

FASTA headers may contain limited organism names but lack standardized metadata.

5. References and Publications

GenBank files include scientific references related to the sequence.

Example:

REFERENCE   1
AUTHORS     Smith et al.
TITLE       Complete mitochondrial genome analysis

These references allow researchers to:

verify experimental evidence
trace sequence origin
consult published studies

FASTA files do not contain publication or citation information.

Conclusion

GenBank is more informative than FASTA because it combines sequence data with rich biological annotations, genomic features, metadata, and literature references.

This makes GenBank highly useful for:

genome annotation
gene extraction
protein analysis
mutation studies
comparative genomics

In contrast, FASTA is mainly suitable for storing and sharing raw sequence data.

# ***Exercise 6 — Entrez Search Design***

Design Entrez search queries for:


Bos taurus mitochondrial genome
Capra hircus cytochrome b gene
Ovis aries complete mitochondrial genome
A gene of your research interest

In [10]:
# Exercise 6 — Entrez Search Design

from Bio import Entrez
from Bio import SeqIO

# Set your email for NCBI Entrez access
Entrez.email = "legessetsion8@gmail.com"

print("Entrez email set to:", Entrez.email)

# ---------------------------------------------------
# Example Search:
# Bos taurus mitochondrial genome
# ---------------------------------------------------

handle = Entrez.esearch(
    db="nucleotide",
    term="Bos taurus mitochondrion complete genome",
    retmax=5
)

results = Entrez.read(handle)
handle.close()

print("\nNCBI ID List:")
print(results["IdList"])

# ---------------------------------------------------
# Fetch FASTA records
# ---------------------------------------------------

id_list = results["IdList"]

handle = Entrez.efetch(
    db="nucleotide",
    id=id_list,
    rettype="fasta",
    retmode="text"
)

records = list(SeqIO.parse(handle, "fasta"))
handle.close()

# Display first sequence information
print("\nFirst Record Information")
print("Description:", records[0].description)
print("First 100 Bases:")
print(records[0].seq[:100])

# ---------------------------------------------------
# Designed Entrez Search Queries
# ---------------------------------------------------

query1 = '"Bos taurus"[Organism] AND complete mitochondrial genome'

query2 = '"Capra hircus"[Organism] AND cytochrome b[Gene]'

query3 = '"Ovis aries"[Organism] AND complete mitochondrial genome'

query4 = '"Homo sapiens"[Organism] AND BRCA1[Gene]'

# Print queries
print("\nDesigned Entrez Search Queries")

print("\n1. Bos taurus mitochondrial genome")
print(query1)

print("\n2. Capra hircus cytochrome b gene")
print(query2)

print("\n3. Ovis aries complete mitochondrial genome")
print(query3)

print("\n4. Research interest gene (BRCA1)")
print(query4)

Entrez email set to: legessetsion8@gmail.com

NCBI ID List:
['3264155392', '3156478204', '2991366831', '2991366817', '2991366803']

First Record Information
Description: PX725379.1 UNVERIFIED: Bos taurus isolate XBBO024 mitochondrion sequence
First 100 Bases:
CCCATGCTCACACATAACTGTGCTGTCATACATTTGGTATTTTTTTATTTTGGGGGATGCTTGGACTCAGCTATGGCCGTCAAAGGCCCTGACCCGGAGC

Designed Entrez Search Queries

1. Bos taurus mitochondrial genome
"Bos taurus"[Organism] AND complete mitochondrial genome

2. Capra hircus cytochrome b gene
"Capra hircus"[Organism] AND cytochrome b[Gene]

3. Ovis aries complete mitochondrial genome
"Ovis aries"[Organism] AND complete mitochondrial genome

4. Research interest gene (BRCA1)
"Homo sapiens"[Organism] AND BRCA1[Gene]


# ***Exercise 7 — Mutation Interpretation***

In [11]:
from Bio import Align

# Define sequences
seq1 = "ATGCCCTAG"
seq2 = "ATGCGCTAG"

# Create aligner object
aligner = Align.PairwiseAligner()

# Perform alignment
alignments = aligner.align(seq1, seq2)

# Print alignment information
print("Number of alignments:", len(alignments))
print("Best alignment score:", alignments[0].score)

# Print best alignment
print("\nBest Alignment:")
print(alignments[0])

# ---------------------------------------------------
# Identify sequence differences
# ---------------------------------------------------

print("\nDifferences Found:")

for i in range(len(seq1)):
    if seq1[i] != seq2[i]:
        print(f"Position {i+1}: {seq1[i]} → {seq2[i]}")

Number of alignments: 5
Best alignment score: 8.0

Best Alignment:
target            0 ATGC-CCTAG  9
                  0 ||||-|-||| 10
query             0 ATGCGC-TAG  9


Differences Found:
Position 5: C → G


# ***Exercise 8 — Alignment Statistics***

For each column in the alignment, count how many different bases are present. Which columns are variable?

In [13]:
from Bio import AlignIO
from collections import Counter

# Create alignment FASTA file
alignment_fasta = """>sample1
ATGCGTACGTAG
>sample2
ATGCGTTCGTAG
>sample3
ATGCGTACGCAG
"""

with open("example_alignment.fasta", "w") as handle:
    handle.write(alignment_fasta)

# Read alignment
alignment = AlignIO.read("example_alignment.fasta", "fasta")

# Print alignment info
print(alignment)
print("Number of sequences:", len(alignment))
print("Alignment length:", alignment.get_alignment_length())

print("\n" + "="*50)
print("Exercise 8 — Alignment Statistics")
print("="*50)

# Convert alignment into sequence list
seqs = [str(record.seq) for record in alignment]

# Alignment length
alignment_length = len(seqs[0])

# Analyze each column
for i in range(alignment_length):

    # Get bases in column
    column = [seq[i] for seq in seqs]

    # Count bases
    counts = Counter(column)

    print(f"\nPosition {i+1}")
    print("Bases:", dict(counts))

    # Identify variable columns
    if len(counts) > 1:
        print("Variable column")
    else:
        print("Conserved column")

Alignment with 3 rows and 12 columns
ATGCGTACGTAG sample1
ATGCGTTCGTAG sample2
ATGCGTACGCAG sample3
Number of sequences: 3
Alignment length: 12

Exercise 8 — Alignment Statistics

Position 1
Bases: {'A': 3}
Conserved column

Position 2
Bases: {'T': 3}
Conserved column

Position 3
Bases: {'G': 3}
Conserved column

Position 4
Bases: {'C': 3}
Conserved column

Position 5
Bases: {'G': 3}
Conserved column

Position 6
Bases: {'T': 3}
Conserved column

Position 7
Bases: {'A': 2, 'T': 1}
Variable column

Position 8
Bases: {'C': 3}
Conserved column

Position 9
Bases: {'G': 3}
Conserved column

Position 10
Bases: {'T': 2, 'C': 1}
Variable column

Position 11
Bases: {'A': 3}
Conserved column

Position 12
Bases: {'G': 3}
Conserved column


## Exercise 9 — Tree Interpretation

Based on the tree above, which two samples are more closely related? Why?

The two samples that are more closely related are:

- sample1
- sample2

## Why?

In the phylogenetic tree:

((sample1:0.1,sample2:0.2):0.3,sample3:0.4);

sample1 and sample2 are grouped together in the same branch before joining with sample3.

This means they share a more recent common ancestor.

Samples that branch together in a phylogenetic tree are considered more closely related.

sample3 joins the tree later, meaning it is more distantly related.

## Conclusion

sample1 and sample2 are the most closely related samples because they cluster together and share the nearest common ancestor.

# ***Exercise 10 — Improve the Pipeline***

In [16]:
from Bio import SeqIO
import pandas as pd

# Function to calculate GC content
def calculate_gc(sequence):
    sequence = str(sequence).upper()

    gc_count = sequence.count("G") + sequence.count("C")

    return (gc_count / len(sequence)) * 100


# Function to calculate AT content
def calculate_at(sequence):
    sequence = str(sequence).upper()

    at_count = sequence.count("A") + sequence.count("T")

    return (at_count / len(sequence)) * 100


# Function to classify GC category
def gc_category(gc):

    if gc >= 60:
        return "High GC"

    elif gc >= 40:
        return "Medium GC"

    else:
        return "Low GC"


# Improved pipeline function
def analyze_fasta(input_fasta, output_csv):

    results = []

    for record in SeqIO.parse(input_fasta, "fasta"):

        seq = record.seq
        seq_string = str(seq).upper()

        # GC and AT content
        gc_content = calculate_gc(seq)
        at_content = calculate_at(seq)

        # Reverse complement
        reverse_comp = str(seq.reverse_complement())

        # Check for ambiguous bases
        contains_ambiguous = "N" in seq_string

        # Translation
        translated = ""
        stop_codons = 0

        if seq_string.startswith("ATG") and len(seq_string) % 3 == 0:

            translated = str(seq.translate(to_stop=False))

            # Count stop codons (*)
            stop_codons = translated.count("*")

        # Save results
        results.append({
            "id": record.id,
            "description": record.description,
            "length": len(seq),
            "gc_content": round(gc_content, 2),
            "at_content": round(at_content, 2),
            "gc_category": gc_category(gc_content),
            "starts_with_ATG": seq_string.startswith("ATG"),
            "length_multiple_of_3": len(seq_string) % 3 == 0,
            "contains_N": contains_ambiguous,
            "reverse_complement": reverse_comp,
            "translated_sequence": translated,
            "stop_codon_count": stop_codons
        })

    # Create DataFrame
    df = pd.DataFrame(results)

    # Save CSV
    df.to_csv(output_csv, index=False)

    return df


# Run pipeline
pipeline_results = analyze_fasta(
    "example_sequences.fasta",
    "sequence_summary.csv"
)

# Display results
print(pipeline_results)

     id              description  length  gc_content  at_content gc_category  \
0  seq1  seq1 Example sequence 1      20        50.0        50.0   Medium GC   
1  seq2  seq2 Example sequence 2      21        61.9        38.1     High GC   
2  seq3  seq3 Example sequence 3      21         0.0       100.0      Low GC   

   starts_with_ATG  length_multiple_of_3  contains_N     reverse_complement  \
0             True                 False       False   CTAGCTAGCTACGTACGCAT   
1             True                  True       False  CCCGGGTTTAAACCCGGGCAT   
2            False                  True       False  TATATATATATATATATATAT   

  translated_sequence  stop_codon_count  
0                                     0  
1             MPGFKPG                 0  
2                                     0  


# Final Challenge Project — Option C
# Mitochondrial Genome Comparison

## Objective

Retrieve mitochondrial genome sequences from multiple species using Biopython and compare:

- sequence length
- GC content

This project demonstrates:
- Entrez database searching
- FASTA retrieval
- sequence analysis
- summary table generation

---



In [30]:
from Bio import Entrez
from Bio import SeqIO
import pandas as pd

# Set email
Entrez.email = "legessetsion8@gmail.com"

# Species list
species_list = [
    "Bos taurus",
    "Capra hircus",
    "Ovis aries",
    "Sus scrofa"
]

# GC content function
def calculate_gc(sequence):

    sequence = str(sequence).upper()

    gc_count = sequence.count("G") + sequence.count("C")

    return (gc_count / len(sequence)) * 100


# Store results
results = []

# Process each species
for species in species_list:

    print(f"Searching for {species} mitochondrial genome...")

    # Corrected search query
    handle = Entrez.esearch(
        db="nucleotide",
        term=f'"{species}"[Organism] AND mitochondrion[Title] AND complete genome',
        retmax=1
    )

    search_results = Entrez.read(handle)
    handle.close()

    id_list = search_results["IdList"]

    # Skip if no records found
    if len(id_list) == 0:
        print("No records found.")
        continue

    print("NCBI ID:", id_list[0])

    # Fetch FASTA sequence
    handle = Entrez.efetch(
        db="nucleotide",
        id=id_list[0],
        rettype="fasta",
        retmode="text"
    )

    record = SeqIO.read(handle, "fasta")
    handle.close()

    seq = record.seq

    # Calculate statistics
    length = len(seq)
    gc_content = calculate_gc(seq)

    a_count = seq.count("A")
    t_count = seq.count("T")
    g_count = seq.count("G")
    c_count = seq.count("C")

    # Save results
    results.append({
        "Species": species,
        "Sequence_Length": length,
        "GC_Content": round(gc_content, 2),
        "A_Count": a_count,
        "T_Count": t_count,
        "G_Count": g_count,
        "C_Count": c_count
    })

# Create dataframe
df = pd.DataFrame(results)

# Display results
print("\nMitochondrial Genome Summary")
print(df)

# Save CSV
df.to_csv("mitochondrial_genome_summary.csv", index=False)

print("\nSummary saved to mitochondrial_genome_summary.csv")

Searching for Bos taurus mitochondrial genome...
NCBI ID: 3264155392
Searching for Capra hircus mitochondrial genome...
NCBI ID: 3336904142
Searching for Ovis aries mitochondrial genome...
NCBI ID: 3264155393
Searching for Sus scrofa mitochondrial genome...
NCBI ID: 3139154378

Mitochondrial Genome Summary
        Species  Sequence_Length  GC_Content  A_Count  T_Count  G_Count  \
0    Bos taurus            16308       39.08     5416     4397     2171   
1  Capra hircus            16639       39.14     5581     4545     2180   
2    Ovis aries            16432       38.61     5472     4450     2122   
3    Sus scrofa            16739       39.54     5805     4316     2229   

   C_Count  
0     4202  
1     4333  
2     4223  
3     4389  

Summary saved to mitochondrial_genome_summary.csv


# Conclusion

Mitochondrial genome sequences from four livestock species were successfully retrieved from the NCBI nucleotide database using Biopython Entrez.

The analysis compared:
- sequence length
- GC content
- nucleotide composition

Results showed that all mitochondrial genomes had similar lengths, approximately 16,000–17,000 base pairs, which is typical for animal mitochondrial DNA.

GC content values were also relatively similar across species, ranging from approximately 38% to 40%.

This project demonstrated how Biopython can be used for:
- online database searching
- sequence retrieval
- FASTA parsing
- genomic sequence analysis
- summary table generation

The workflow is useful in comparative genomics, livestock genetics, and bioinformatics research.